File: ruby/lib/sorpresa_salircarcel.rb
module Civitas
    class SorpresaSalirCarcel < Sorpresa
        public_class_method :new
        def initialize(mazo)
            super("Salir de carcel")
            @mazo=mazo
        end

        def aplicarAJugador(actual, todos)
            if jugadorCorrecto(actual, todos)
                informe(actual, todos)
                tiene=false
                for i in 0...todos.size do
                    tiene=todos[i].tieneSalvoConducto
                    if tiene
                        break
                    end
                end
                if !tiene
                    todos[actual].obtenerSalvoconducto(self)
                    self.salirDelMazo
                end
            end
        end

        def salirDelMazo
            @mazo.inhabilitarCartaEspecial(self)
        end

        def usada
            @mazo.habilitarCartaEspecial(self)
        end
    end
end

File: Civitas_gui/src/civitas/SorpresaSalirCarcel.java
package civitas;
import java.util.ArrayList;

public class SorpresaSalirCarcel extends Sorpresa {
    private MazoSorpresas mazo;

    SorpresaSalirCarcel(MazoSorpresas mazo) {
        super("Salir de la carcel");
        this.mazo = mazo;
    }

    public void aplicarAJugador(int actual, ArrayList<Jugador> todos) {
        if (jugadorCorrecto(actual, todos)){
            informe(actual,todos);
            boolean alguienTieneSalvoConducto = false;
            for (int i = 0; i < todos.size() && !alguienTieneSalvoConducto; i++) {
                if (todos.get(i).tieneSalvoconducto()) {
                    alguienTieneSalvoConducto = true;
                }

                if (!alguienTieneSalvoConducto) {
                    todos.get(actual).obtenerSalvoconducto(this);
                    salirDelMazo();
                }
            }
        }
    }

    void salirDelMazo() {
        mazo.inhabilitarCartaEspecial(this);
    }

    void usada() {
        mazo.habilitarCartaEspecial(this);
    }
}